### ISSUE TITLE
Company doesn't work in Emacs 29: "backend company-files error "Invalid function""

### ISSUE THREAD

**Issue by kuba-orlik** [company-mode/company-mode#1381]
#### Output of the command `M-x company-diag`

The command doesn't run and just throws a message saying:

```
Company: An error occurred in auto-begin
Company: backend company-files error "Invalid function: ("\"\\(\\(?:\\.\\{1,2\\}/\\|~/\\|/\\)[^\"
]*\\)" "'\\(\\(?:\\.\\{1,2\\}/\\|~/\\|/\\)[^'
]*\\)" "\\(?:[ 	=[]\\|^\\)\\(\\(?:\\.\\{1,2\\}/\\|~/\\|/\\)[^ 	
]*\\)")" with args (prefix)
```

#### Describe the issue
Short description of the problem.
Copy/paste the exact error message: you can find it in the **\*Messages\*** buffer.

#### Steps to reproduce

1. `emacs -Q`
2. execute this lisp code:

```
(setq straight-repository-branch "develop")
(defvar bootstrap-version)
(let ((bootstrap-file
       (expand-file-name "straight/repos/straight.el/bootstrap.el" user-emacs-directory))
      (bootstrap-version 5))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
         "https://raw.githubusercontent.com/raxod502/straight.el/develop/install.el"
         'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))
(straight-use-package 'company)
(add-hook 'after-init-hook 'global-company-mode)
```

#### The error backtrace

```
Debugger entered--Lisp error: (invalid-function ("\"\\(\\(?:\\.\\{1,2\\}/\\|~/\\|/\\)[^\"\n]*\\)" "'\\(\\(?:\\.\\{1,2\\}/\\|~/\\|/\\)[^'\n]*\\)" "\\(?:[ \11=[]\\|^\\)\\(\\(?:\\.\\{1,2\\}/\\|~/\\|/\\)[^ \11\n]*\\)"))
  ("\"\\(\\(?:\\.\\{1,2\\}/\\|~/\\|/\\)[^\"\n]*\\)" "'\\(\\(?:\\.\\{1,2\\}/\\|~/\\|/\\)[^'\n]*\\)" "\\(?:[ \11=[]\\|^\\)\\(\\(?:\\.\\{1,2\\}/\\|~/\\|/\\)[^ \11\n]*\\)")(nil)
  company-files--grab-existing-name()
  company-files(prefix)
  apply(company-files prefix)
  company-call-backend-raw(prefix)
  company--force-sync(company-call-backend-raw (prefix) company-files)
  company-call-backend(prefix)
  company--begin-new()
  company--perform()
  company-auto-begin()
  company-idle-begin(#<buffer carbon.org> #<window 3 on carbon.org> 304 3238)
  apply(company-idle-begin (#<buffer carbon.org> #<window 3 on carbon.org> 304 3238))
  timer-event-handler([t 25643 10523 96922 nil company-idle-begin (#<buffer carbon.org> #<window 3 on carbon.org> 304 3238) nil 691000 nil])
```

#### Additional context
I'm running Emacs 29.0.60

**Comment by dgutov** [comment]
Hi! Thanks for the report. I haven't been able to reproduce this, however.

@kuba-orlik Do you use the very latest version from emacs-29?

@technician77 Do you also install using `straight`?

**Comment by technician77** [comment]
No, just used package.el. It is the latest pre-compiled windows version from the gnu alpha ftp. (2023-03-28)

**Comment by dgutov** [comment]
Hmm, I have tried installing from MELPA using a build from 2023-03-27, and the very latest version from emacs-29. These -- using package.el.

And I have tried the scenario from the first message using `straight`. Still don't see the error.

**Comment by dgutov** [comment]
What happens if you remove the package and then install it again? Do you see any relevant warnings/errors during byte-compilation? I'm assuming the ones mentioning `company-files` would be most relevant.

If you're using `straight`, try clearing the cache in `~/.emacs.d/straight` before reinstalling.

**Comment by technician77** [comment]
Deleting and reinstalling company did not change the situation.

Below is hopefully all information you need. I did not deinstall all other packages I just went with -Q option, hope this is sufficient. I have some other packages that relate to company like consult-company and company-quickhelp but since they are not "required" in the config below it should make no difference.

Steps to reproduce:
1. Install GNU Emacs 29.0.60 (build 1, x86_64-w64-mingw32) of 2023-03-28
3. Install company via packagel.el
4. Paste in scratch:
```
(require 'package)
(package-initialize)
(require 'company)
(company-mode)
```
5.M-x eval-buffer
6.Type ( in scratch.

My Result:
```
Company: An error occurred in auto-begin
Company: backend company-files error "Invalid function: ("\"\\(\\(?:\\.\\{1,2\\}/\\|~/\\|[a-zA-Z]:/\\)[^\"
]*\\)" "'\\(\\(?:\\.\\{1,2\\}/\\|~/\\|[a-zA-Z]:/\\)[^'
]*\\)" "\\(?:[ 	=[]\\|^\\)\\(\\(?:\\.\\{1,2\\}/\\|~/\\|[a-zA-Z]:/\\)[^ 	
]*\\)")" with args (prefix)
```

**Comment by willemml** [comment]
I have the same issue (using emacs.git/emacs-29 (I also tried using emacs.git/master (emacs-30))) using the the latest git commit of company-mode. Currently I have disabled the company-files backend to avoid the repeated error pop-ups. To do this I simply added `(delete 'company-files company-backends)` to my Emacs configuration. This does however mean I do not get file name completion.

**Comment by dgutov** [comment]
@technician77 

Ok, thanks.

> No errors during compiling / async-native-compiling.

This is the weird part: the error looks like the macro `cl-dolist` was not recognized by the byte compiler (i.e. `cl-lib` was not loaded), but the first thing `company-files.el` does is `(require 'cl-lib)`.

And usually, when a missing macro is an issue, a warning is issued during byte-compilation (when the package is installed) which looks very much like your error.

You can try visiting `company-files.el` and calling `M-x eval-buffer`. Does that make the error go away in the current session?

**Comment by dgutov** [comment]
@willemml I guess that confirms that it's not just the issue of the mingw build. Or of a specific commit.

I'm not sure why I'm not seeing it, though. Even though I've tried different builds of Emacs (with native compilation and without), and an empty config dir (using `HOME=/tmp path/to/emacs`).

Do you build your Emacs yourself? You can try `make extraclean` followed by `make bootstrap` and then uninstalling/reinstalling the package. Just `company-mode` -- no other ones needed.

**Comment by willemml** [comment]
It may be of interest that it also consistently happens when I am editing a commented line of code and that I also use eglot.

In regards to compiling I currently use a nix derivation to build it (so a clean build from source without any cache every time I recompile). I will see if the issue persists with an isolated installation of emacs and company-mode.

**Comment by dgutov** [comment]
> It may be of interest that it also consistently happens when I am editing a commented line of code and that I also use eglot.

That shouldn't really matter.

The problem hints at a miscompilation, which happens when the package is being installed (and the .elc files are generated). It's either a bug in the package (which makes it odd since I don't see it), or an issue with a particular version of Emacs (revision or type of build), which would be even less likely, but the first option has been all but ruled out by now, I think.

> In regards to compiling I currently use a nix derivation to build it (so a clean build from source without any cache every time I recompile). I will see if the issue persists with an isolated installation of emacs and company-mode.

Thanks.

Before that, you can try and experiment with `M-x eval-buffer` inside `company-files.el` like I mentioned previously, or deleting `company-files.elc` (the compiled artefact) from the package's directory and restarting Emacs (`(load "company-files.el")` should be an alternative to that). And seeing which of these help. Maybe that gives some ideas.

**Comment by willemml** [comment]
So I cloned the repo and then ran `emacs -Q company-files.el` and ran `M-x eval-buffer`. This produced the following output:
```
Compiling /nix/store/iw6p9ha94pqqxl8kizarb6g561mmlibd-emacs29/share/emacs/29.0.60/lisp/emacs-lisp/cl-loaddefs.el.gz...
uncompressing cl-loaddefs.el.gz...
uncompressing cl-loaddefs.el.gz...done
Compilation finished.
Compiling /nix/store/gzc0302mad99ymbni27bx51j1ilvl6si-emacs-packages-deps/share/emacs/site-lisp/elpa/company-20230404.140/company.el...
Compilation finished.
```
I then created a new buffer and enabled `emacs-lisp-mode` and ran `(global-company-mode)`.
I then cleared the buffer and type as single `(` and received no errors, I also tried setting a variable to a string containing a filename (which company quickly provided autocomplete options for me).

In short it seems to work if I only have `company-mode`.

**Comment by dgutov** [comment]
> Also when I enabled global-company-mode the rest of the company elisp files were compiled (from my package cache in the nix store) but company-files.el was never compiled (according to Emacs logs). Output of this is as follows:

This looks like the "native compilation" log which happens on-demand (in builds where native compilation is enabled), compiling only files that are loaded. Any chance that your setting with `(delete 'company-files company-backends)` was also present in that session? Alternatively, it might have been not compiled because the package was loaded from source and not from `.elc` file.

**Comment by dgutov** [comment]
I don't generally run with native compilation, so I'm probably missing some nuance here -- e.g. I don't see these compilation logs in my `-Q` session at all. Even though that Emacs was compiled with native compilation enabled.

**Comment by dgutov** [comment]
And about this part:

> So I cloned the repo and then ran emacs -Q company-files.el and ran M-x eval-buffer

How come that session already had the directory with all `company` package files inside `load-path`? When I do that (with `company` installed using package.el), I just get an error:

```el
eval-buffer: Cannot open load file: No such file or directory, company
```

Guess that's Nix altering load-path by default somehow?

**Comment by willemml** [comment]
> This looks like the "native compilation" log which happens on-demand (in builds where native compilation is enabled), compiling only files that are loaded. Any chance that your setting with (delete 'company-files company-backends) was also present in that session?

I am certain that it was not.

> Guess that's Nix altering load-path by default somehow?

Pretty much, all the packages that I use get installed and symlinked directly into emacs' site lisp and then a wrapper loads them.

I just tried running without that wrapper, and I redid the steps with eval-buffer in the cloned repo, this time it did not compile *any* of the files from company-mode (I did however have to run `(add-to-list 'load-path "/Users/willem/cmtest/company-mode")` to avoid the error you mentioned).

On top of that I no longer have the original error, so it is probably something to do with how it gets compiled.

**Comment by dgutov** [comment]
> I am certain that it was not (especially given that completions provided by company-files were available, unless I misunderstand its purpose).

All right, thanks for verifying.

> On top of that I no longer have the original error, so it is probably something to do with how it gets compiled.

It would be great if you could pinpoint two similar scenarios, both not using your full configuration, where one of them triggers the problem, and the other does not. That would help us find the cause of the issue, or at least minimize the reproduction recipe and (most probably) report it upstream. Subtle nativecomp bugs is something we most probably want to take care of before 29's release.

Is the use of the "wrapper" the sole/main cause? Given that @technician77 is using a mingw build, I'm guessing the full Nix environment should not be necessary to hit the problem.

**Comment by willemml** [comment]
It actually seems that I can no longer reproduce the error inside or outside of my config at all now. I am hesitant to say this just yet but that's how it seems. I did check that the `company-files` backend is still enabled and it is, filepath completions are still working.

So the things I have tried are (without the wrapper and with `emacs -Q`) downloading company from ELPA, downloading from source and manually native compiling all of the files.
Then with the wrapper I have tried company from ELPA and company from the git source repo.

**Comment by dgutov** [comment]
Hmm, one of the things that might be hindering reproduction is the caching of native compilation artefacts: they usually reside inside `~/.emacs.d/eln-cache/<emacs version>-<some hash>/`. The eln files are named after the package name with some content-based hashes, I think.

`emacs -Q` might skip the full config, but it can still look in the same dir to see whether a native compiled version of the package already exists.

So the thing to try is `rm ~/.emacs.d/eln-cache/29.0.60-8adbdaec/company*` between reproduction attempts (update the name of the parent directory as needed if your hash is different).

**Comment by willemml** [comment]
I tried after clearing my compilation cache, but company-mode appears to still be working just fine, I can no longer reproduce the error.

**Comment by dgutov** [comment]
Okay, then it's possible that miscompilation happened once due to some rare circumstances, possibly using some previous build of Emacs too.

@kuba-orlik @technician77 Try deleting the company .eln files from your `eln-cache` directory too, does it fix the problem?

**Comment by technician77** [comment]
Deleting the eln cache did change nothing, but deleting the elc files in the company package directory did.

**Comment by dgutov** [comment]
Okay. But when you reinstall the package afterwards, does the problem come back?

**Comment by technician77** [comment]
Yes, I also found out I only needed to delete the company-files elc.

Company is now up again, but now I found out my company-quickhelp stopped working too. Now the funny part. Based on the results so far I thought, just delete the company-quickhelp elc file and "tada" it worked again. 
`Error running timer 'company-quickhelp--show': (invalid-function 0)`

I'm not a coder, so I'm not sure how the compiling process exactly works, but could it be related to my MSYS2 environment? I think on windows native compiling only works when MSYS2 is installed and some development packages like gcc. Well, there might be other ways, but that's what I use.

**Comment by technician77** [comment]
Emacs 29.0.60 (build 1, x86_64-w64-mingw32) of 2023-03-10 is good. Reinstalled company and company-help. No errors anymore . Had nothing to do with MSYS2. I will reinstall all packages. Maybe the bad compilation process broke also other packages, which errors I have yet to discover. Maybe some emacs dev should take a look at this?

**Comment by dgutov** [comment]
> Yes, I also found out I only needed to delete the company-files elc.

Deleting elc files is one thing -- that is expected to fix things, but the proper installed state of the packages is to have the compiled .elc files as well. Because byte-compilation speeds up Lisp execution by about 2x, last time I measured that.

But I see you reinstalled the packages and the problem went away, good.

This probably means that the current build you are using is fine.

> Maybe the bad compilation process broke also other packages, which errors I have yet to discover. Maybe some emacs dev should take a look at this?

That's indeed something to bring up with the devs, but ideally we'd have a step-by-step repro.

Do I understand it right that if you revert to the previous build that you were using and reinstall `company` and `company-help`, the problem comes back? That would confirm the above explanation. Do you know the exact Git hash it was built from?

Have you tried any more recent builds of Emacs? I'm not sure where you get yours, but https://alpha.gnu.org/gnu/emacs/pretest/windows/emacs-29/ has some build from 2023-04-05 ([emacs-29.0.60-5e1953.zip](https://alpha.gnu.org/gnu/emacs/pretest/windows/emacs-29/emacs-29.0.60-5e1953.zip)). That's probably not for MSYS2, though.

**Comment by technician77** [comment]
It seems the emacs devs already fixed it with the latest built. So, i'm good now. Thanks for analyzing. I would never have come up with the idea, that the package is o.k. and the problem is a broken compilation process inside emacs.

No problems:	                      			emacs-29.0.60_1.zip		2023-03-10 13:25	152M	 	 
Missing DLLs+Compilation broken: 	emacs-29-64a2b0d36fe.zip	2023-03-25 12:18	137M	 
Compilation broken:		               	emacs-29-edc460e3b6c.zip	2023-03-28 00:57	152M
No problems:			                        emacs-29.0.60-5e1953.zip	2023-04-05 21:34	152M

**Comment by dgutov** [comment]
Excellent, thanks for verifying.

Looking at the revision range, this was likely related to https://debbugs.gnu.org/61917 (breakage and the fixes fit the time periods).

To summarize for anyone else who comes across this report with the same problem: you need to update to a more recent build of Emacs, reinstall the package (and any other package with similar symptoms), and probably also clear the directory `eln-cache`.
